Support polynomial division for mask

juliuskunze opened the pull request:
Add support for multivariate polynomial division on polymorphic sizes without remainder, allowing `mask` of

- `jnp.reshape` with -1 size and
- `lax.slice` for polymorphic stride for sizes  `poly * stride`, i. e. `(n^2+2n)//n = n+2`

Also clean up `Poly` class, improve error messages.

JuliusKunze commented:
@mattjj This small and new PR is now ready to review. (;

gnecula commented on jax/interpreters/masking.py:
```diff
@@ -239,6 +245,12 @@ def evaluate(self, env):
     return len(self) == 1 and next(iter(self)).degree == 0
 
+  @property
+  def _leading_term(self) -> Tuple['Mon', int]:
```
What is the significance of the leading_term? Is it the constant if one is present?

JuliusKunze replied:
It's needed by polynomial division to determine the next term to divide.

gnecula commented on jax/numpy/lax_numpy.py:
```diff
@@ -1135,7 +1135,7 @@ def check(size):
   newshape = [check(size) for size in newshape] if iterable else check(newshape)
   newsize = _prod((newshape,) if type(newshape) is Poly else newshape)
-  if newsize < 0:
+  if np.any(np.equal(newshape, -1)):
```
I think this is a better check than before. I am wondering though if other negative dimensions will be caught reliably somewhere else.

JuliusKunze replied:
Comment a few lines above said: "# other errors, like having more than one -1, are caught downstream".
In any case, the change does not worsen error handling.

gnecula commented on jax/interpreters/masking.py:
```diff
@@ -107,77 +107,93 @@ def eval_poly_shape(shape, values_dict):
   return Poly({Mon(): p})
 
-def is_polymorphic(shape: Sequence[Union[int, 'Poly']]):
+def is_polymorphic(shape: Sequence['Size']):
```
I saw the definition later. If you move it up, you won't need to put in in quotes

JuliusKunze replied:
Had the same thought, but one can't move it up, since it requires Poly to be defined already.

gnecula commented on jax/lax/lax.py:
```diff
@@ -3599,9 +3599,9 @@ def _slice_shape_rule(operand, *, start_indices, limit_indices, strides):
-  return tuple(result_shape)
+  diff = np.subtract(limit_indices, start_indices)
+  # Not np.divmod since Poly.__rdivmod__ is ignored by NumPy, breaks poly stride
+  return tuple(q + np.where(r, 1, 0) for q, r in map(divmod, diff, strides))
```
I was initially confused by the use of `np.where`, thinking that it is going to break in presence of polynomials. But `r` is always a scalar, right? Then perhaps `(r > 0)` is cleaner than `np.where(r, 1, 0)`?

JuliusKunze replied:
Good point, updated to `r > 0`!

gnecula commented on jax/interpreters/masking.py:
```diff
@@ -253,33 +265,43 @@ def mul(coeff, mon):
   def degree(self):
     return sum(self.values())
 
+  def __floordiv__(self, divisor: 'Mon') -> Optional['Mon']:
```
Since you clearly figured out the representation of these monomials, do you mind adding a few comments to this class at the top?

JuliusKunze replied:
Ok, done!

gnecula commented on jax/interpreters/masking.py:
```diff
@@ -107,77 +107,93 @@ def eval_poly_shape(shape, values_dict):
+    """Multivariate polynomial division, returning quotient and remainder."""
+    dmon, dcount = divisor._leading_term
+    dividend, quotient, remainder = self, _ensure_poly(0), _ensure_poly(0)
+    while dividend != 0:
```
I think it would help if you wrote down the loop invariant that dividend = divisor * quotient + remainder. 

I am trying to convince myself that this is "complete". E.g., if the leading term of the dividend and the divisor are not the same, then the monomial division will return None, and the entire leading term of the dividend is added to the remainder. I guess I am not sure what means the leading term. If it were the one with the highest degree, I could see how this terminates. But otherwise, why does it terminate?

JuliusKunze replied:
@gnecula I described the invariant in the code. 

The leading terms is the highest degree term that comes first lexicographically. (The monomials are ordered in [graded reverse lexicographical order](https://en.wikipedia.org/wiki/Monomial_order#Graded_reverse_lexicographic_order): First by total degree and then lexicographically to break ties.)

gnecula reviewed (commented):
This is a welcome change, thank you! I have a few comments, perhaps I did not fully understand the details.

mattjj commented:
Multivariate polynomial division in general requires a doubly exponential time algorithm; this is what [Grobner bases](https://en.wikipedia.org/wiki/Gr%C3%B6bner_basis) are for. I don't think that's a direction we want to go.

mattjj commented:
Thanks for continuing to try new ideas @JuliusKunze ! To say a bit more, it's intentional that we don't support `lax.slice` with polymorphic stride: in general `a // b` is not a polynomial for general polynomials, and a key desideratum is that the language of polymorphic shapes be closed and tractable to decide equality (polynomial equality can be checked in BPP).

Our current plan with strides is to support only division of the form `(c * poly + k) // c == poly + k // c` for integers `c` and `k`. If we can enlarge that to a bigger family of tractable polynomial divisions then we should! But the key question is what that family is, as we know general multivariate polynomial division is intractable.

gnecula commented:
I have made some comments in a related design doc that is not yet external, so I repeat them here. 

I think it is important to have a clear mental model of what is supported and what is not in terms of polynomial division, but we should think how to enlarge the class of supported division operation. It would be unsatisfying to see an error that '6x' cannot be divided by 'x'. 

I propose at least to add to the supported class the dual of the one @mattjj pointed out: `(c * poly) // poly ==  c`. 

I am also wondering how often people will only need univariate polynomials. In a use case I am thinking about, we use one shape variable to be standing for the batch dimension. Perhaps it makes sense to also support division of univariate polynomials. 

Yet another case that I think may occur frequently is when the polynomials are just monomials (no `+`). I am not sure, and it would be best to get some experience using this before we decide what further exceptions to add.

JuliusKunze commented on jax/interpreters/masking.py:
```diff
@@ -107,77 +107,93 @@ def eval_poly_shape(shape, values_dict):
+  """Polynomial with integer coefficients for polymorphic shapes."""
 
-  def __init__(self, coeffs):
+  def __init__(self, coeffs: Dict['Mon', int]):
```
I did intentionally not do that, for now, to minimize the PR diff.

JuliusKunze commented on jax/interpreters/masking.py:
```diff
@@ -107,77 +107,93 @@ def eval_poly_shape(shape, values_dict):
+    """
+    Floor division with remainder (divmod) generalized to polynomials.
+    To allow ensuring consistency with integer divmod,
+    the polynomial division remainder divided by the divisor must be constant.
```
Clarified the description!

JuliusKunze commented:
@mattjj perfect division (`(c * poly) // poly == c` as described by @gnecula) is exactly the class this PR adds. +1 for adding it to the spec since it's useful for the listed cases and doesn't require Gröbner bases.

@gnecula The cases of univariate polynomials and polynomials with only one term are already covered by that extension, or did you have another use case in mind?

gnecula commented:
This LGTM but I know that @mattjj has been working on updating this part of the code. He should be the one to give this PR the final verdict.

gnecula commented on jax/interpreters/masking.py:
```diff
@@ -253,33 +262,53 @@ def mul(coeff, mon):
 
-
 class Mon(dict):
+  """Represents a multivariante monomial, such as n^3 * m."""
```
Typo: multivariate

gnecula commented on jax/interpreters/masking.py:
```diff
@@ -253,33 +262,53 @@ def mul(coeff, mon):
 
-  def __lt__(self, other):
-    # sort by total degree, then lexicographically on indets
+  def __lt__(self, other: 'Mon'):
```
I find this strange, and hard to understand. IIUC, this is not the numeric comparison, just a sorting order for monomials based on their representation. 

The "-self.degree" along with ">" used in an implementation of "__lt__" is confusing.  

Is this needed only for _leading term? If so, wouldn't it be clearer if in that function we define this as sorting key without overloading __lt__?

JuliusKunze replied:
@gnecula I think what you suggest is not equivalent, see https://en.wikipedia.org/wiki/Monomial_order#Graded_reverse_lexicographic_order

I replaced the direction here since `x^2<x` (how it was before) is unintuitive.
Semantically, various other orders could be used without breaking poly division. I chose this order to allow creating string representations following `PolyTest.test_parse_spec` just by sorting.

gnecula commented:
@mattjj I think that this PR is worth merging. It does enable a common idiomatic use of `reshape(-1)`.

gnecula commented on jax/interpreters/masking.py:
```diff
@@ -107,77 +107,90 @@ def eval_poly_shape(shape, values_dict):
+      qmon = mon // dmon
+      qcount, rcount = divmod(count, dcount)
+      q = Poly({qmon: qcount}) if qmon else _ensure_poly(0)
+      r = Poly({mon: rcount if qmon else count})
```
I still find this code very hard to understand. 

Does it even make sense to continue if qmon is None or if rcount is not 0? The remainder will be non-constant. An exception has to be made for the case when the divisor is a constant. If we abort early, the code is easier to understand: we only consider the case when the leading term of the divisor divides the leading term of the dividend.

JuliusKunze replied:
@gnecula Good point, simplified it.

juliuskunze committed:
Support polynomial division for mask

Add support for multivariate polynomial division on polymorphic sizes without remainder, allowing `mask` of

- `jnp.reshape` with -1 size and
- `lax.slice` for polymorphic stride for sizes  `poly * stride`, i. e. `(n^2+2n)//n = n+2`

Also clean up `Poly` class, improve error messages.

gnecula commented:
@JuliusKunze I am sorry this took so long, I will merge it. 

Unfortunately, there are a few pytype failures:
```
File "third_party/py/jax/interpreters/masking.py", line 161, in __floordiv__: Built-in function divmod was called with the wrong arguments [wrong-arg-types]
         Expected: (x: int, ...)
  Actually passed: (x: Poly, ...)
File "third_party/py/jax/interpreters/masking.py", line 165, in __mod__: Built-in function divmod was called with the wrong arguments [wrong-arg-types]
         Expected: (x: int, ...)
  Actually passed: (x: Poly, ...)
File "third_party/py/jax/interpreters/masking.py", line 194, in __rdivmod__: Built-in function divmod was called with the wrong arguments [wrong-arg-types]
         Expected: (x: int, ...)
  Actually passed: (x: Poly, ...)
```

These come from pytype, which we use inside Google.

gnecula commented:
I will take care of fixing these errors on import.

mattjj commented:
LGTM. Thanks for following up on this, George and @JuliusKunze.

I agree about supporting special cases of division that are easy. Thanks for clarifying the scope of these changes!

mattjj reviewed (approved):
LGTM. I started (but have yet to finish) a refactor that will likely affect some of the implementation details here, but not the functionality, which sounds good!
